expr/types.go

```diff
@@ -0,0 +1,91 @@
+// argument types. to let functions describe their inputs and outputs
+package expr
+
+type arg interface {
```

Comment:
It's definitely possible, following on from our discussion about making this a standalone library there may well be scenarios where someone using it wants to provide additional functions specific to their use-case, or to override the implementation of particular function(s).